numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -3855,12 +3878,32 @@ compare_chararrays(PyObject *NPY_UNUSED(dummy), PyObject *args, PyObject *kwds)
         Py_DECREF(newarr);
         return NULL;
     }
-    if (PyArray_ISSTRING(newarr) && PyArray_ISSTRING(newoth)) {
+    int is_user_string_newarr = _is_user_defined_string_array(newarr);
+    int is_user_string_newoth = _is_user_defined_string_array(newoth);
+    if ((PyArray_ISSTRING(newarr) || (is_user_string_newarr == 1)) &&
+        (PyArray_ISSTRING(newoth) || (is_user_string_newoth == 1)))
+    {
         res = _umath_strings_richcompare(newarr, newoth, cmp_op, rstrip != 0);
```

Comment:
I don't buy that you can go into this function if you have a user-string.  Yes, it probably works for your ASCIIDType, but it won't work for a string dtype that contains a pointer.

My suggestion would be the following: Just dump `rstrip=True` (raise an error), and for `rstrip=False`, you can call the normal richcompare if you like.  I suspect we could just deprecate `rstrip` anyway...  (but if someone needs it, implement it as rstrip + comparison at some point)

Reply:
I think I'm just going to back out my changes to `compare_chararrays` since we can implement all of the comparisons using ufuncs anyway.